nakedret.go

```diff
@@ -23,41 +29,57 @@ func init() {
 	build.Default.UseAllFiles = true
 }
 
-func usage() {
-	log.Printf("Usage of %s:\n", os.Args[0])
-	log.Printf("\nnakedret [flags] # runs on package in current directory\n")
-	log.Printf("\nnakedret [flags] [packages]\n")
-	log.Printf("Flags:\n")
-	flag.PrintDefaults()
+func NakedReturnAnalyzer(defaultLines uint) *analysis.Analyzer {
+	nakedRet := &NakedReturnRunner{}
+	flags := flag.NewFlagSet("nakedret", flag.ExitOnError)
+	flags.UintVar(&nakedRet.MaxLength, "l", defaultLines, "maximum number of lines for a naked return function")
+	var analyzer = &analysis.Analyzer{
+		Name:     "nakedret",
+		Doc:      "Checks that functions with naked returns are not longer than a maximum size (can be zero).",
+		Run:      nakedRet.run,
+		Flags:    *flags,
+		Requires: []*analysis.Analyzer{inspect.Analyzer},
+	}
+	return analyzer
+}
+
+type NakedReturnRunner struct {
+	MaxLength uint
+}
+
+func (n *NakedReturnRunner) run(pass *analysis.Pass) (interface{}, error) {
+	inspector := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+
+	nodeFilter := []ast.Node{ // filter needed nodes: visit only them
+		(*ast.FuncDecl)(nil),
+		(*ast.FuncLit)(nil),
+		(*ast.ReturnStmt)(nil),
+	}
+	retVis := &returnsVisitor{
+		pass:      pass,
+		f:         pass.Fset,
+		maxLength: n.MaxLength,
+	}
+	inspector.Nodes(nodeFilter, retVis.NodesVisit)
+	return nil, nil
 }
 
 type returnsVisitor struct {
+	pass      *analysis.Pass
 	f         *token.FileSet
 	maxLength uint
-	root      *returnsVisitor
-	found     bool
 
+	stack []funcInfo
```

Comment:
Maybe it's better to use`functions` as the slice's name? Because `stack` is unclear.

Reply:
agree, also thought about this.

updated